assayist/processor/container_go_analyzer.py

```diff
@@ -0,0 +1,139 @@
+
+    # goversion executable path.
+    GOVERSION = 'goversion'
+
+    # Backvendor executable path.
+    BACKVENDOR = 'backvendor'
+
+    # Output format tempate for backvendor.
+    BACKVENDOR_TEMPLATE = '\t{{.Ver}}\t{{.Repo}}\t{{.Rev}}'
+
+    # Types of identified source modules
+    UPSTREAM = 'upstream'
+    VENDORED = 'embedded_source_locations'
+
+    def run(self):
+        """Start the container Go analyzer."""
+        build_info = self.read_metadata_file(self.BUILD_FILE)
+        build_id = build_info['id']
+        if not self.is_container_build(build_info):
+            log.info(f'Skipping build {build_id} because the build is not a container')
+            return
+
+        # This container's build is assumed to exist since it is
+        # created by the main analyzer.
+        build = content.Build.nodes.get(id_=build_id)
+        source_location = build.source_location
+        if not source_location:
+            log.error('Missing source location for container build')
+            return
+
+        self._process_source_code(source_location)
+        self._claim_go_executables()
+
+    def _process_source_code(self, source_location):
+        """Run backvendor on the source code.
+
+        :param SourceLocation source_location: local source code DB node
+        """
+        # Find the source code.
+        srcdir = os.path.join(self.input_dir, self.SOURCE_DIR)
+        options = ['-template', self.BACKVENDOR_TEMPLATE]
+        cmd = [self.BACKVENDOR] + options + ['.']
+        bv = subprocess.Popen(cmd, cwd=srcdir,
+                              stdout=subprocess.PIPE,
+                              stderr=subprocess.PIPE)
+        (stdout, stderr) = bv.communicate()
+        returncode = bv.wait()
+        if returncode:
+            raise RuntimeError(f'The command "{" ".join(cmd)}" failed with: {stderr}')
+
+        # There may be many top-level repositories, but we are only
+        # allowed a single 'upstream' relationship type. Only report
+        # the first top-level repository as "the upstream" one.
+        seen_toplevel = False
+
+        # Parse the output from backvendor.
+        for line in stdout.splitlines():
+            fields = line.split('\t')
+            if len(fields) != 4:
+                log.error(f'invalid backvendor output: {line}')
+                continue
+
+            mod, ver, repo, rev = fields
+
+            # The module field begins with an asterisk for a top-level module
+            srctype = self.VENDORED
+            if mod.startswith('*'):
+                mod = mod[1:]
+                if not seen_toplevel:
+                    srctype = self.UPSTREAM
+                    seen_toplevel = True
+
+            self._process_go_module(source_location, srctype,
+                                    mod, ver, repo, rev)
+
+    def _process_go_module(self, source_location, srctype,
+                           mod, ver, repo, rev):
```

Comment:
`rev` doesn't seem to be used below, should be part of the version string if `ver` does not exist?

Reply:
Ah, seems like Analyzer._create_or_update_source_location() takes a url parameter which is "the url and possibly commit hash of the source location".

Should I join them with "#"?

Reply:
Yes. `ver` is the version, and we use that as the canonical version. I've made a change to join the `rev` to the URL using '#'.